neuro_flow/batch_executor.py

```diff
@@ -62,23 +65,166 @@ def parse(cls, raw: str) -> "ExecutorData":
         return cls(**data)
 
 
+_T = TypeVar("_T")
+
+
+class GraphEmb(Generic[_T]):
+    """TopologicalSorter that allows to embed sub graphs in nodes
+
+    Instance of this graph is constructed by provided simple graph
+    specification. After that, additional graphs can be embedded
+    to nodes of this (or previously embedded) graph. Each node
+    has can address using FullID, where each component is names
+    node in some graphs. For example, node id ('top', 'node')
+    Is a node of graph is embedded to node 'top' of the main top
+    level graph.
+
+    Node is considered ready when it is ready inside its graph
+    and the parent node is ready. Top level graph is ready by
+    default.
+
+    Each graph can be annotated with some metadata.
+    """
+
+    def __init__(self, graph: Mapping[str, Iterable[str]], meta: _T):
+        topo = graphlib.TopologicalSorter(graph)
+        self._topos: Dict[FullID, "graphlib.TopologicalSorter[str]"] = {(): topo}
+        self._metas: Dict[FullID, _T] = {(): meta}
+        topo.prepare()
+        self._ready: Set[FullID] = set()
+        self._done: Set[FullID] = set()
+        self._embeds_done: Set[FullID] = set()
+        self._process_ready(())
+
+    @property
+    def is_active(self) -> bool:
+        return bool(self._ready)
+
+    def get_ready(self) -> AbstractSet[FullID]:
+        return set(self._ready)
+
+    def get_ready_with_meta(self) -> Iterable[Tuple[FullID, _T]]:
+        for ready in set(self._ready):
+            yield ready, self._metas[ready[:-1]]
+
+    def get_meta(self, node: FullID) -> _T:
```

Comment:
`meta` is actually a node's value, isn't it?

Reply:
It is metadata attached to **graph** that contains given node. This is because we offen need to access ctx

Reply:
Ok. The `meta` name is not the best choice maybe but it works well enough.
We can change it later.